Libraries/LibURL/Pattern/ConstructorStringParser.cpp

```diff
@@ -0,0 +1,703 @@
+PatternErrorOr<void> ConstructorStringParser::compute_protocol_matches_a_special_scheme_flag()
+{
+    // FIXME: Implement this.
+    return {};
+}
+
+Optional<String> const& ConstructorStringParser::result_for_active_state() const
+{
+    switch (m_state) {
+    case State::Protocol:
+        return m_result.protocol;
+    case State::Username:
+        return m_result.username;
+    case State::Password:
+        return m_result.password;
+    case State::Hostname:
+        return m_result.hostname;
+    case State::Port:
+        return m_result.port;
+    case State::Pathname:
+        return m_result.pathname;
+    case State::Search:
+        return m_result.search;
+    case State::Hash:
+        return m_result.hash;
+    case State::Initial:
+    case State::Authority:
+    case State::Done:
+        break;
+    }
+    VERIFY_NOT_REACHED();
+}
+
+void ConstructorStringParser::set_result_for_active_state(Optional<String> value)
+{
+    switch (m_state) {
+    case State::Protocol:
+        m_result.protocol = move(value);
+        break;
+    case State::Username:
+        m_result.username = move(value);
+        break;
+    case State::Password:
+        m_result.password = move(value);
+        break;
+    case State::Hostname:
+        m_result.hostname = move(value);
+        break;
+    case State::Port:
+        m_result.port = move(value);
+        break;
+    case State::Pathname:
+        m_result.pathname = move(value);
+        break;
+    case State::Search:
+        m_result.search = move(value);
+        break;
+    case State::Hash:
+        m_result.hash = move(value);
+        break;
+    case State::Initial:
+    case State::Authority:
+    case State::Done:
+        VERIFY_NOT_REACHED();
+    }
+}
+// https://urlpattern.spec.whatwg.org/#change-state
+void ConstructorStringParser::change_state(State new_state, u32 skip)
+{
+    // 1. If parser’s state is not "init", not "authority", and not "done", then set parser’s result[parser’s state] to
+    //    the result of running make a component string given parser.
+    if (m_state != State::Initial && m_state != State::Authority && m_state != State::Done)
+        set_result_for_active_state(make_a_component_string());
+
+    // 2. If parser’s state is not "init" and new state is not "done", then:
+    if (m_state != State::Initial && new_state != State::Done) {
+        // 1. If parser’s state is "protocol", "authority", "username", or "password"; new state is "port", "pathname",
+        //    "search", or "hash"; and parser’s result["hostname"] does not exist, then set parser’s result["hostname"]
+        //    to the empty string.
+        if ((m_state == State::Protocol || m_state == State::Authority || m_state == State::Username || m_state == State::Password)
+            && (new_state == State::Port || new_state == State::Pathname || new_state == State::Search || new_state == State::Hash)
```

Comment:
`first_is_one_of` makes long chains of comparisons like this a bit more readable IMO:
```c++
first_is_one_of(m_state, State::Protocol, State::Authority, State::Username, State::Password)
```